src/layers/domain/core/entity.py

```diff
@@ -18,3 +23,6 @@ def __eq__(self, other):
 
     def __ne__(self, other):
         return not self.__eq__(other)
+
+    def as_reference(self) -> Reference:
```

Comment:
the whole `Reference` thing comes across as a code smell: why are converting objects to string references rather referencing the actual objects? Python by default will treat objects as references, and if the issue is about namespacing ids then we can just modify the `__hash__` function above to include the type.__name__